### PR TITLE
add phpunit 6

### PR DESCRIPTION
**Pull request by samsonasik**
or, we can just use phpunit 6 since we use php7

### CODE REVIEW

File: system/Test/CIUnitTestCase.php [hunk 1]
```diff
@@ -36,13 +36,17 @@
  * @filesource
  */
 
-use PHPUnit_Framework_TestCase;
+use PHPUnit\Framework\TestCase;
 use CodeIgniter\Log\TestLogger;
 
+if (! class_exists(TestCase::class)) {
+    class_alias(\PHPUnit_Framework_TestCase::class, TestCase::class);
+}
```
Comment (samsonasik) [inline system/Test/CIUnitTestCase.php]: upgrading to phpunit 6 will allow to not doing class_alias hack. Just use new `PHPUnit\Framework\TestCase`

**Comment by samsonasik** [comment]
I am closing and re-open to restart build in favor latest phpunit 6.

**Comment by kenjis** [comment]
@samsonasik At the moment, CI4 supports PHPUnit 5.3: https://github.com/bcit-ci/CodeIgniter4/blob/develop/composer.json#L23

@lonnieezell Does CI4 drop PHPUnit 5.x and before? I think PHPUnit (major) version up is very fast. So once CI4 is officially released, we should fix the version of PHPUnit for some long time.
But CI4 is now under development and PHP7 is required. So I think no problem if CI4 drops PHPUnit 5.x.

**Comment by lonnieezell** [comment]
Sorry, hadn't had a chance to respond on this thread yet. Yes, I think it's fine to fix it to 6 from the initial look at what's changed. And, as @kenjis said, that should get us in a good place where we can leave it in the 6.x branch for a while. 

So, yes, I'm okay with dropping support for PHPunit 5.x. Heck, there's part of me that doesn't think it even needs to be in our composer.json file, just list it as a requirement. Is there really a need to have a new copy with every site? People can install globally, download the .phar, or add it to their own...

**Comment by samsonasik** [comment]
I've updated phpunit to v6

**Comment by samsonasik** [comment]
phpunit should be in `vendor` in each project imo not globally

**Comment by samsonasik** [comment]
Got 2 errors:

```
There were 2 failures:

1) CodeIgniter\Helpers\URLHelperTest::testPreviousURLUsesRefererIfNeeded
PHP Fatal error:  Method PHPUnit\Framework\ExpectationFailedException::__toString() must not throw an exception, caught ErrorException: Undefined index: SCRIPT_NAME in /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php on line 0
PHP Stack trace:
PHP   1. {main}() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/phpunit:0
PHP   2. PHPUnit\TextUI\Command::main() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/phpunit:53
PHP   3. PHPUnit\TextUI\Command->run() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/Command.php:135
PHP   4. PHPUnit\TextUI\TestRunner->doRun() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/Command.php:206
PHP   5. PHPUnit\TextUI\ResultPrinter->printResult() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:520
PHP   6. PHPUnit\TextUI\ResultPrinter->printFailures() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:199
PHP   7. PHPUnit\TextUI\ResultPrinter->printDefects() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:300
PHP   8. PHPUnit\TextUI\ResultPrinter->printDefect() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:243
PHP   9. PHPUnit\TextUI\ResultPrinter->printDefectTrace() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:256

Fatal error: Method PHPUnit\Framework\ExpectationFailedException::__toString() must not throw an exception, caught ErrorException: Undefined index: SCRIPT_NAME in /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php on line 0

Call Stack:
    0.0005     358832   1. {main}() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/phpunit:0
    0.0204     710736   2. PHPUnit\TextUI\Command::main() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/phpunit:53
    0.0204     710848   3. PHPUnit\TextUI\Command->run() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/Command.php:135
    1.1114   10473672   4. PHPUnit\TextUI\TestRunner->doRun() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/Command.php:206
   36.5843   38731592   5. PHPUnit\TextUI\ResultPrinter->printResult() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:520
   36.5844   38731592   6. PHPUnit\TextUI\ResultPrinter->printFailures() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:199
   36.5844   38731592   7. PHPUnit\TextUI\ResultPrinter->printDefects() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:300
   36.5844   38731592   8. PHPUnit\TextUI\ResultPrinter->printDefect() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:243
   36.5845   38731592   9. PHPUnit\TextUI\ResultPrinter->printDefectTrace() /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php:256


An uncaught Exception was encountered

Type:        ErrorException
Message:     Method PHPUnit\Framework\ExpectationFailedException::__toString() must not throw an exception, caught ErrorException: Undefined index: SCRIPT_NAME
Filename:    /Users/samsonasik/www/CodeIgniter4/vendor/phpunit/phpunit/src/TextUI/ResultPrinter.php
Line Number: 0

	Backtrace:
```

**Comment by kenjis** [comment]
> phpunit should be in vendor in each project imo not globally

I think so too. Every project has its own PHPUnit version. The global PHPUnit version may be different from the version of a project. In that case, after all, we have to install PHPUnit in the project.

PHPUnit in composer.json clearly states its version requirement.
If users don't need it, they can remove it.

**Comment by kenjis** [comment]
@samsonasik You should set `$_SERVER['SCRIPT_NAME']` in the test code.

The error is because the PHPUnit 6.0 change:
* Global and super-global variables are no longer backed up before and restored after each test by default

In other words, the default of PHPUnit 6.0 is now `backupGlobals="false"`.

We could set `backupGlobals="true"`, but it is memory hungry. And if our code is clean, we don't depend global variables. I think we should not depend on `backupGlobals`.

**Comment by samsonasik** [comment]
It seems not "only" that, see https://travis-ci.org/bcit-ci/CodeIgniter4/builds/198630214 that still error which I was added `$_SERVER['SCRIPT_NAME']`, so I removed it at https://github.com/bcit-ci/CodeIgniter4/pull/391/commits/684f4e3edb4d11d6e32e83ad811f8f3fac660873

**Comment by kenjis** [comment]
@samsonasik You're right.

How about this?
```diff
--- a/tests/system/Helpers/URLHelperTest.php
+++ b/tests/system/Helpers/URLHelperTest.php
@@ -15,6 +15,14 @@ class URLHelperTest extends \CIUnitTestCase
        Services::reset();
    }
 
+ public function tearDown()
+ {
+     unset($_SERVER['HTTP_HOST']);
+     unset($_SERVER['REQUEST_URI']);
+     unset($_SERVER['HTTP_REFERER']);
+     unset($_SESSION);
+ }
+
    //--------------------------------------------------------------------
    // Test site_url
 
```

**Comment by samsonasik** [comment]
@kenjis Thank you. that still got error: 
```php
1) CodeIgniter\Router\RouteCollectionTest::testCLI
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'here' => '\there'
 )
```
I set `backupGlobals="true"` in phpunit.xml.dist . It seems a workaround for it ;)

**Comment by samsonasik** [comment]
travis green ;) https://travis-ci.org/bcit-ci/CodeIgniter4/builds/199040575

**Comment by lonnieezell** [comment]
Very strange. any idea why the code coverage dropped 11% with these changes? I'm not see what would cause that, but that's a huge change.

**Comment by samsonasik** [comment]
I've rebased against develop. If the coverage still dropped, it may be related with php-code-coverage bug against phpunit 6

**Comment by lonnieezell** [comment]
@samsonasik do you have a link with details on a code coverage bug or is that just a possibility?

**Comment by samsonasik** [comment]
just possibility for now

**Comment by kenjis** [comment]
@lonnieezell @samsonasik 

See:
* http://kenjis.github.io/tmp/ci4-coverage-phpunit6/coverage-5.3.5/Commands/index.html
* http://kenjis.github.io/tmp/ci4-coverage-phpunit6/coverage-6.0.5/Commands/index.html

It seems PHPUnit6's coverage is more accurate. Because we have no tests for Commands:
https://github.com/bcit-ci/CodeIgniter4/tree/develop/tests/system

**Comment by samsonasik** [comment]
@kenjis thank you for the verify 👍

**Comment by kenjis** [comment]
I got it. 

The total line is the following:
* PHPUnit 5.3.5: 4836 / 7855
* PHPUnit 6.0.5: 4828 / 9453

The green lines are 4836 and 4828. Only dropped 8 lines. But the total lines are very different.

**Comment by kenjis** [comment]
The 8 lines are here:
* http://kenjis.github.io/tmp/ci4-coverage-phpunit6/coverage-5.3.5/CLI/CLI.php.html - 40
* http://kenjis.github.io/tmp/ci4-coverage-phpunit6/coverage-6.0.5/CLI/CLI.php.html - 32

**Comment by kenjis** [comment]
The difference is the `wait()`:
* http://kenjis.github.io/tmp/ci4-coverage-phpunit6/coverage-5.3.5/CLI/CLI.php.html#341
* http://kenjis.github.io/tmp/ci4-coverage-phpunit6/coverage-6.0.5/CLI/CLI.php.html#341

**Comment by kenjis** [comment]
I'm not sure why `wait()` is differnet, but my conclusion is the coverage by PHPUnit 5.3 was not more accurate than 6.0's. The coverage by PHPUnit 6.0 is no problem.

**Comment by lonnieezell** [comment]
Ah, ok, that makes sense. Thanks for looking into that @kenjis, I've been slammed with work and a kid that's getting sick so my available time currently isn't much. 

So, do you guys both think we're good to merge in this change?

**Comment by samsonasik** [comment]
I think it is good to merge ;)

**Comment by kenjis** [comment]
Me too.

Take care of your kid!

**Comment by kenjis** [comment]
After merging, I will look into <https://github.com/bcit-ci/CodeIgniter4/pull/391#issuecomment-277859053>.

I'm going to make `backupGlobals` false. Because it is the default of PHPUnit6 and I think it is better.

**Comment by lonnieezell** [comment]
@kenjis Fixing the backupGlobals thing sounds like a good long-term fix, anyway. Thanks. 

I'll merge this in, then.

**Pull request merged**